packages/contracts/contracts/L1/AssetManager.sol

```diff
@@ -0,0 +1,1428 @@
+ */
+abstract contract AssetManager is IERC721Receiver {
+    using Math for uint256;
+    using Uint128Math for uint128;
+    using BalancedWeightTree for BalancedWeightTree.Tree;
+
+    /**
+     * @notice Represents the asset information of the vault of a validator.
+     *
+     * @custom:field totalKro                   Total amount of KRO in the vault.
+     * @custom:field totalKroShares             Total shares for KRO delegation in the vault.
+     * @custom:field totalKgh                   Total number of KGH in the vault.
+     * @custom:field totalKroInKgh              Total amount of KRO which KGHs in the vault have.
+     * @custom:field totalKghShares             Total shares for KGH delegation in the vault.
+     * @custom:field validatorKro               Amount of KRO that the validator self-delegated.
+     * @custom:field totalPendingAssets         Total pending KRO for undelegation.
+     * @custom:field totalPendingBoostedRewards Total pending boosted rewards in KRO for undelegation.
+     * @custom:field totalPendingKroShares      Total pending KRO shares for undelegation.
+     * @custom:field totalPendingKghShares      Total pending KGH shares for undelegation.
+     */
+    struct Asset {
+        uint128 totalKro;
+        uint128 totalKroShares;
+        uint128 totalKgh;
+        uint128 totalKroInKgh;
+        uint128 totalKghShares;
+        uint128 validatorKro;
+        uint128 totalPendingAssets;
+        uint128 totalPendingBoostedRewards;
+        uint128 totalPendingKroShares;
+        uint128 totalPendingKghShares;
+    }
+
+    /**
+     * @notice Constructs the reward information of the vault of a validator.
+     *
+     * @custom:field boostedReward                 Cumulated boosted reward for KGH delegators in the vault.
+     * @custom:field validatorRewardKro            Cumulated reward for the validator.
+     * @custom:field commissionRateNumerator       Numerator of the commission rate.
+     * @custom:field commissionRateMaxChangeRange  Range that a commission rate can be changed to.
+     * @custom:field lastCommissionRateChangedTime Last timestamp when the commission rate was changed.
+     */
+    struct Reward {
+        uint128 boostedReward;
+        uint128 validatorRewardKro;
+        uint8 commissionRateNumerator;
+        uint8 commissionRateMaxChangeRange;
+        uint128 lastCommissionRateChangedTime;
+    }
+
+    /**
+     * @notice Constructs the delegator of KRO in the vault of a validator.
+     *
+     * @custom:field shares                 Amount of shares for KRO delegation.
+     * @custom:field undelegateRequestTimes Timestamps of undelegation requests.
+     * @custom:field undelegateRequests     Pending shares at timestamps undelegations are requested.
+     */
+    struct KroDelegator {
+        uint128 shares;
+        uint256[] undelegateRequestTimes;
+        mapping(uint256 => uint128) undelegateRequests;
```

Comment:
Why don't we rename to `pendingShares`? The term `Request` is quite ambiguous to me.

Reply:
I agree with that, could you check the following namings are sufficient?

- `undelegateRequests` (at `KroDelegator`) -> `pendingKroShares` 
- `undelegateRequests` (at `KghDelegator`) -> `pendingKghShares`
- `undelegateRequestedIds` -> `pendingTokenIds`

Reply:
All looks good. But how about `pendingKghIds`?

BTW, for abbreviation, capitals are used in golang, but not in contract. If so, let's use `kgh`.
(e.g. kgh -> KGH, Id -> ID)